pkg/txn/storage/txn/log_tail.go

```diff
@@ -0,0 +1,191 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package txnstorage
+
+import (
+	"math"
+
+	"github.com/matrixorigin/matrixone/pkg/container/batch"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	apipb "github.com/matrixorigin/matrixone/pkg/pb/api"
+	"github.com/matrixorigin/matrixone/pkg/pb/timestamp"
+	"github.com/matrixorigin/matrixone/pkg/pb/txn"
+	txnengine "github.com/matrixorigin/matrixone/pkg/vm/engine/txn"
+)
+
+type LogTailEntry = apipb.Entry
+
+func (m *MemHandler) HandleGetLogTail(meta txn.TxnMeta, req txnengine.GetLogTailReq, resp *txnengine.GetLogTailResp) (err error) {
+
+	// get table
+	m.tables.Lock()
+	table, ok := m.tables.Map[req.TableID]
+	if !ok {
+		m.tables.Unlock()
+		resp.ErrRelationNotFound.ID = req.TableID
+		return nil
+	}
+	m.tables.Unlock()
+
+	// time range
+	from := timestamp.Timestamp{
+		PhysicalTime: math.MinInt64,
+	}
+	if req.Request.CnHave != nil {
+		from = *req.Request.CnHave
+	}
+	to := timestamp.Timestamp{
+		PhysicalTime: math.MaxInt64,
+	}
+	if req.Request.CnWant != nil {
+		to = *req.Request.CnWant
+	}
+	fromTime := Time{
+		Timestamp: from,
+	}
+	toTime := Time{
+		Timestamp: to,
+	}
+
+	// tx
+	tx := m.getTx(meta)
+
+	// attributes
+	rows, err := m.attributes.IndexRows(tx, AnyKey{
+		index_RelationID,
+		Text(req.TableID),
+	})
+	if err != nil {
+		return err
+	}
+	attrsMap := make(map[string]*AttributeRow)
+	insertNames := make([]string, 0, len(rows))
+	deleteNames := make([]string, 0, len(rows))
+	for _, row := range rows {
+		attrsMap[row.Name] = row
+		insertNames = append(insertNames, row.Name)
+		if row.Primary || row.IsRowId {
+			deleteNames = append(deleteNames, row.Name)
+		}
+	}
+
+	// batches
+	insertBatch := batch.New(false, insertNames)
+	for i, name := range insertNames {
+		insertBatch.Vecs[i] = vector.New(attrsMap[name].Type)
+	}
+	deleteBatch := batch.New(false, deleteNames)
+	for i, name := range deleteNames {
+		deleteBatch.Vecs[i] = vector.New(attrsMap[name].Type)
+	}
+
+	// iter
+	iter := table.rows.Copy().Iter()
+	defer iter.Release()
+	for ok := iter.First(); ok; ok = iter.Next() {
+		physicalRow := iter.Item()
+
+		values := physicalRow.Values
+		values.RLock()
+		for i := len(values.Values) - 1; i >= 0; i-- {
+			value := values.Values[i]
+
+			if value.LockTx != nil &&
+				value.LockTx.State.Load() == Committed &&
+				value.LockTime.After(fromTime) &&
+				value.LockTime.Before(toTime) {
+				// committed delete
+				namedRow := &NamedAnyRow{
+					Row:      value.Value,
+					AttrsMap: attrsMap,
+				}
+				if err := appendNamedRow(tx, m.mheap, deleteBatch, namedRow); err != nil {
```

Comment:
When is the memory released? I don't see the code to release it

Reply:
fixed